zkml/src/quantization/strategy.rs

```
@@ -217,105 +153,103 @@ impl InferenceTracker {
-                    }
-                }
-            })
-            .collect::<Vec<Layer<Element>>>();
-        Ok((
-            Model::<Element>::new_from(quantized_layers, input_not_padded_shape, input_shape),
-            md.build(),
-        ))
+        quantize_model::<AbsoluteMax>(model, (), input_scaling_factor)
+    }
+}
+
+fn quantize_model<S: ScalingStrategy>(
+    model: Model<f32>,
+    data: S::AuxData,
+    input_scaling: Vec<ScalingFactor>,
+) -> anyhow::Result<(Model<Element>, ModelMetadata)>
+where
+    Dense<f32>: QuantizeOp<S, QuantizedOp = Dense<Element>>,
+    Convolution<f32>: QuantizeOp<S, QuantizedOp = Convolution<Element>>,
+{
+    let input_shapes = model.input_shapes();
+    let input_not_padded_shapes = model.unpadded_input_shapes();
+    let mut md = MetadataBuilder::new(input_scaling);
+    // 2. Create the requant layers from the infered data
+    let mut requant_layers = vec![];
+    let mut catch_err: Result<()> = Ok(());
+    let nodes = model
+        .into_forward_iterator()
+        .map(|(node_id, node)| {
+            let input_scaling = md.compute_input_scaling(&node.inputs)?;
+            let quantized_out =
+                quantize_op::<S, _>(node.operation, &data, node_id, &input_scaling)?;
+            md.set_layers_scaling(node_id, quantized_out.output_scalings, input_scaling);
+            if let Some(requant) = quantized_out.requant_layer {
+                requant_layers.push((node_id, requant));
+            }
+            let quantized_node =
+                Node::new_with_outputs(node.inputs, quantized_out.quanzited_op, node.outputs);
+            Ok((node_id, quantized_node))
+        })
+        .map_while(|n| {
+            if n.is_err() {
+                catch_err = Err(n.unwrap_err());
+                None
+            } else {
+                Some(n.unwrap())
+            }
+        });
+    let mut model = Model::new_from_shapes(input_not_padded_shapes, input_shapes, nodes);
+    catch_err?;
```

I understand the concerns about not collecting first while keeping the iterator argument on the model creation... but failing early is often, if not always, a better choice. 
Would it be acceptable to give a hashmap to Model ? i dont see it as an horror show since anyway it's just another way to represent the same information `(node_id, node)` which is imo even cleaner. 
Said differently, we can keep the same memory consumption at the cost of changing the container of the info `(node_id, node)` and yet being able to panic early in case of problems.